### PR TITLE
Add timezone to datetime and when renderers

### PR DESCRIPTION
**Pull request by anhuynh**
# Overview
Add timezone into `datetime` and `when` renderers

## Summary
Provide a general summary of the issue addressed in the title above

## Screenshots or recordings
screen shot 2018-07-30 at 10 02 23 am
screen shot 2018-07-30 at 10 24 47 am

## Checklist
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] I have evaluated if the _README.md_ documentation needs to be updated
* [x] I have evaluated if the _/tests/dummy/_ app needs to be modified
* [ ] I have evaluated if DocBlock headers needed to be added or updated
* [ ] I have verified that lint and tests pass locally with my changes
* [ ] If a fork of a dependent package had to be made to address the issue this PR closes:
  * [ ] I noted in the fork's _README.md_ the reason the fork was created
  * [ ] I have opened an upstream issue detailing what was deficient about the dependency
  * [ ] I have opened an upstream PR addressing this deficiency
  * [ ] I have opened an issue in this repository to track this PR and schedule the removal of the usage of the fork

# Semver

**This project uses [semver](http://semver.org), please check the scope of this PR:**

- [ ] #none#
- [ ] #patch#
- [x] #minor#
- [ ] #major#

# CHANGELOG

* Add timezone options to `datetime` and `when` renderers

### CODE REVIEW

File: addon/components/inputs/abstract-input.js [hunk 1]
```diff
@@ -156,6 +156,12 @@ export default Component.extend(HookMixin, PropTypeMixin, {
     return required && valueEmpty
   },
 
+  @readOnly
+  @computed('cellConfig')
+  alwaysShowRequiredLabel (cellConfig) {
```
Comment (sophypal) [inline addon/components/inputs/abstract-input.js]: Why was this necessary?

File: addon/components/inputs/datetime.js [hunk 2]
```diff
@@ -43,12 +53,14 @@ export default AbstractInput.extend({
   },
 
   @readOnly
-  @computed('defaultToCurrentDateTime', 'value')
-  currentValue (defaultToCurrentDateTime, value) {
-    if (!defaultToCurrentDateTime || value) {
-      return value
+  @computed('defaultToCurrentDateTime', 'timezone', 'storedDateTimeValue')
```
Comment (sophypal) [inline addon/components/inputs/datetime.js]: Why are you moving away from `value`? If consumers decide to pass in a value directly into the `formValue`, this renderer won't update with the correct value. It only seems to initialize with the correct value at initialization, not updates.

File: addon/components/inputs/datetime.js [hunk 3]
```diff
@@ -57,8 +69,43 @@ export default AbstractInput.extend({
...
+     * case that the user selects away from the date-time picker radio button but then returns
+     * @param {Object} value - the users selected date/time moment() object
+     * @returns {undefined}
+     */
+    selectDate (value) {
+      const timezone = this.get('timezone')
+      const datetime = getMomentTimezone(value, timezone, true)
+      const formattedTime = getFormattedTime(datetime, this.get('dateTimeFormat'), timezone)
+
+      this.set('storedDateTimeValue', formattedTime)
```
Comment (sophypal) [inline addon/components/inputs/datetime.js]: This internal state will have the tendency to get out of date with the actual form's value. In future versions of Bunsen, we won't be re-rendering components needlessly so you can't rely on `init` to keep it in sync.

File: addon/components/inputs/when.js [hunk 4]
```diff
@@ -106,6 +108,19 @@ export default AbstractInput.extend({
...
 
+  @readOnly
+  @computed('cellConfig')
+  timezone (cellConfig) {
+    return getWithDefault(cellConfig, 'renderer.timezone', '')
+  },
+
+  @readOnly
+  @computed('storedDateTimeValue', 'timezone')
+  displayStoredDateTimeValue (storedDateTimeValue, timezone) {
```
Comment (sophypal) [inline addon/components/inputs/when.js]: Hmm. I realize now the `storedDateTimeValue` is a artifact of the `when` renderer which you probably just copied. `when` renderer does suffer from the same problems I mentioned.

File: addon/components/inputs/abstract-input.js (hunk 1 above)
Comment (anhuynh) [reply addon/components/inputs/abstract-input.js]: I thought I needed that but actually don't so I'll remove it

File: addon/components/inputs/when.js (hunk 4 above)
Comment (anhuynh) [reply addon/components/inputs/when.js]: Ah yes, after closer inspection, I now see why the `storedDateTimeValue` is needed in the `when` renderer. I'll go back to using `value` in `datetime` but not sure if we can get rid of `storedDateTimeValue` in the `when` renderer

Comment (sophypal) [reply addon/components/inputs/when.js]: Thanks. Stick with datetime for now. We'll address `when` at a future date.

File: addon/components/inputs/datetime.js [hunk 5]
```diff
@@ -57,8 +69,47 @@ export default AbstractInput.extend({
...
+    this._super(...arguments)
+
+    const value = this.get('value')
+    const timezone = this.get('timezone')
+    let currentDateTime = getMomentTimezone(value, timezone)
+
+    if (!this.get('defaultToCurrentDateTime')) {
+      currentDateTime = value
+    } else {
+      currentDateTime = getFormattedTime(currentDateTime, this.get('dateTimeFormat'), timezone)
```
Comment (sophypal) [inline addon/components/inputs/datetime.js]: If I'm reading this correctly, it says if `defaultToCurrentDateTime` is true, then it always overrides the value to the current datetime. I don't think that's what we want. We only want to default it when there is no value.

File: addon/components/inputs/datetime.js [hunk 6]
```diff
@@ -57,8 +69,47 @@ export default AbstractInput.extend({
...
+    }
+
+    this.setProperties({
+      localTimezone: moment().format('Z')
+    })
+
+    if (!isNone(value)) {
```
Comment (sophypal) [inline addon/components/inputs/datetime.js]: You should only call `onChange` if the value is different previously. For example, if the initial value was `null` or `undefined`, you'd want to call `onChange`.  If `value` is `none`, then you'd want to call `onChange`.

File: addon/templates/components/frost-bunsen-input-datetime.hbs [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {{#if (not cellConfig.hideLabel)}}
   <label class={{labelWrapperClassName}}>
     {{renderLabel}}
-    {{#if showRequiredLabel}}
+    {{#if (or showRequiredLabel alwaysShowRequiredLabel)}}
```
Comment (sophypal) [inline addon/templates/components/frost-bunsen-input-datetime.hbs]: You don't need this anymore

File: addon/components/inputs/datetime.js (hunk 6 above)
Comment (anhuynh) [reply addon/components/inputs/datetime.js]: Yeah, that makes sense. I think that this and the stuff you commented on above actually aren't needed. Not quite sure what I was thinking behind this

**Comment by anhuynh** [comment]
@jscharf That looks pretty good. Would have to pass that to UX to see what they think though

File: addon/components/inputs/datetime.js (hunk 6 above)
Comment (sophypal) [reply addon/components/inputs/datetime.js]: That's quite alright. I think the idea was sane. Just needed a little more polish to make sure it worked properly. I would have been okay if you left it but corrected it. As it stands, the component does suffer from not setting the value when the user wants defaults applied. The `currentValue` is only presentational until the first date selection is made.

Leaving it out is fine as well since it's only returning to the same behavior as before.

**Comment by jscharf** [comment]
@anhuynh thanks. I don't want to block this PR if it's good to go, so could you follow up with UX and then do another PR based on their feedback?
